### PR TITLE
Make linalg.lu_factor structured

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67833
* #68188
* #67996
* #68523
* #68522
* #69909
* #69908
* #68185
* #68184
* #68183
* #67014
* __->__ #66934
* #67789
* #66933

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

Differential Revision: [D32684856](https://our.internmc.facebook.com/intern/diff/D32684856)

### CODE REVIEW

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -1651,19 +1652,9 @@ std::tuple<Tensor&, Tensor&, Tensor&> linalg_lu_factor_ex_out(const Tensor& A,
       singleCheckErrors(info.item<int64_t>(), "torch.linalg.lu_factor_ex");
     }
   }
-
```
Comment (bdhirsh) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1654]: Should the bit above here go in the meta function too?
```
  if (check_errors) {
    if (A.dim() > 2) {
      batchCheckErrors(info, "torch.linalg.lu_factor_ex");
    } else {
      singleCheckErrors(info.item<int64_t>(), "torch.linalg.lu_factor_ex");
    }
  }
```

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -237,6 +237,27 @@ TORCH_META_FUNC(triangular_solve)(const Tensor& self, const Tensor& A, bool uppe
...
 
+TORCH_META_FUNC(linalg_lu_factor_ex)(const Tensor& A, bool pivot, bool check_errors) {
+  TORCH_CHECK(A.dim() >= 2, "torch.lu_factor: Expected tensor with 2 or more dimensions. Got size: ", A.sizes(), " instead");
+
+  auto sizes = A.sizes().vec();
+  const auto m = sizes.cend()[-2];
+  const auto n = sizes.cend()[-1];
+
+  // make column major strides for BLAS
+  auto LU_strides = at::native::contiguous_strides(sizes, /*f-contig*=*/true);
```
Comment (bdhirsh) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:248]: Just confirming- it sounds like `f-config` flag ends up doing the same thing as the as the deleted transpose code below?
```
    // Transpose size
    std::iter_swap(req_size.end() - 1, req_size.end() - 2);
    LU.resize_(req_size, MemoryFormat::Contiguous);
    LU.transpose_(-2, -1);  // make 'LU' have Fortran contiguous memory
```

Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:248]: Yes, these two pieces of code are equivalent. This is just a convenient function to use in structured kernels.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:1654]: This checks the error values returned by BLAS / LAPACK. We cannot perform these checks on MetaTensors.

Either that or I did not quite understand what you meant here...

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 2 above)
Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:248]: fwiw, this function was added in the previous PR of this stack: https://github.com/pytorch/pytorch/pull/67789#pullrequestreview-804938218

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (bdhirsh) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:1654]: oh yup you're right- looks like this check involves actual tensor data, not metadata. nvm

**Review by bdhirsh — APPROVED** [review]
LGTM!

**Review by IvanYashchuk — APPROVED** [review]

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684856).

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -1651,19 +1652,9 @@ std::tuple<Tensor&, Tensor&, Tensor&> linalg_lu_factor_ex_out(const Tensor& A,
...
-  auto info = at::empty({0}, A.options().dtype(kInt));
-  at::native::linalg_lu_factor_ex_out(A, pivot, check_errors, LU, pivots, info);
-  return std::make_tuple(std::move(LU), std::move(pivots), std::move(info));
 }
 
-std::tuple<Tensor&, Tensor&> linalg_lu_factor_out(const Tensor& A, bool pivot, Tensor & LU, Tensor & pivots) {
+std::tuple<Tensor&, Tensor&> linalg_lu_factor_out(const Tensor& A, bool pivot, Tensor& LU, Tensor& pivots) {
   auto info = at::empty({0}, A.options().dtype(kInt));
   // We pass check_errors as we want to use lu_factor rather than lu_factor_ex in the errors
   at::linalg_lu_factor_ex_out(LU, pivots, info, A, pivot, /*chech_errors=*/false);
```
Comment (nkaretnikov) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1660]: @Lezcano typo here: `chech_errors`

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684856).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684856).

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 4]
```diff
@@ -237,6 +237,27 @@ TORCH_META_FUNC(triangular_solve)(const Tensor& self, const Tensor& A, bool uppe
   }
 }
 
+TORCH_META_FUNC(linalg_lu_factor_ex)(const Tensor& A, bool pivot, bool check_errors) {
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:240]: Internal errors here:

```
stderr: aten/src/ATen/native/BatchLinearAlgebra.cpp:221:10: error: unused variable 'ndim' [-Werror,-Wunused-variable]
  auto ndim = self_broadcast_size.size();

aten/src/ATen/native/BatchLinearAlgebra.cpp:222:10: error: unused variable 'nrows' [-Werror,-Wunused-variable]
  auto nrows = A.size(-2);
```

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684856).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684856).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684856).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684856).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684856).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684856).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684856).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684856).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684856).

**Pull request closed without merging**